src/Transients.cc

```diff
@@ -275,6 +283,11 @@ Transients::completeWriting(const StoreEntry &e)
 {
     assert(e.hasTransients());
     assert(isWriter(e));
+    // The not freed Transients entry may eventually become 'unattached',
+    // so that it cannot be anchored anymore. Though we handle this situation,
+    // there are some risks the entry metadata becomes stale and, hence, we
+    // should drop such entries. TODO: should we drop this entry immediately to
+    // avoid such risks?
```

Comment:
AFAICT, we are concerned that an idle unattached Transient entry loses its (presumed) ability to become attached to a cache store because another, unrelated transaction purges the corresponding store entry. The Controller must (and now does) prevent such a no-longer-attachable Transient entry from being given to Store clients, but its metadata might become stale, making Controller task more difficult/riskier.

However, Transients::completeWriting() is not the place that deals with idle entries -- Store::Controller::handleIdleEntry() is. _This_ method does not create idle entries. If the corresponding StoreEntry is public, it needs this Transients entry -- we cannot "drop this entry immediately". Otherwise, we should probably rely on handleIdleEntry() being called soon rather than preempt or duplicate its future actions.

Given your existing use cases (that this PR fixes), can you tell why the Transients entry was not freed when it became idle? Or did it become idle _after_ its disk entry was purged, because the Transients entry somehow lost attachment to its disk entry (while remaining busy)?

Reply:
I placed this comment in Transients::completeWriting() because old Squids (pre 4310f8b) used this method to mark the Transients entry for removal:

```
Transients::completeWriting(const StoreEntry &e)
{
    if (e.mem_obj && e.mem_obj->xitTable.index >= 0) {
    [...]
        map->freeEntry(e.mem_obj->xitTable.index); // just marks the locked entry
    [...]
    }
}
```

Once the Transients entry is 'marked', it is not available for the public anymore. Eventually, such 'marked' entry may become unlocked (idle) but this is probably irrelevant here. Perhaps I need to change the comment's 'drop entry' to 'mark entry' for better understanding.

Reply:
Marking the Transients entry in Transients::completeWriting() does not make sense to me: There is nothing wrong with the Transients entry, so why make it inaccessible to new concurrent transactions?

Right now, I see two options for going forward:

1. Remove this comment. The fixes already committed should prevent the known bug from happening again. This is sufficient.
2. Understand what the underlying issue is and add a corresponding TODO to document and address the underlying cause (assuming we do not want to address it in this PR for some valid reason). This is ideal.

It feels like we have or should have enough information to move from the "sufficient" to "ideal" outcome here. That is why I am asking why the Transients entry was not deleted in your use case(s) after it became idle. Did we not call handleIdleEntry() or did handleIdleEntry() decide to keep the idle Transients entry? Why? Or perhaps it never became idle but still lost its lock on the disk (or memory) cache entry? Something else?

And when did the StoreEntry corresponding to the TransientsEntry lost its association with the disk (or memory) store? Is that loss expected/normal?

Somewhere, there is a conflict between our expectations and actual Squid behavior. This bug is the outcome of that conflict. The current fix does not expose which of our basic expectations does not match the code. The fix correctly reacts to a problematic situation, but there may be more we can and should do to either avoid that situation in more cases or adjust our expectations. If we have a use case that can reproduce the original problem, it would be great to use that use case to pin-point the source of that conflict (at least)...

Reply:
> Did we not call handleIdleEntry()

We call handleIdleEntry().

> did handleIdleEntry() decide to keep the idle Transients entry?  Why? 

Yes, it keeps the idle Transients entry.
Transients::handleIdleEntry() calls destroyStoreEntry() and then StoreEntry::destroyMemObject(). Eventually Transients::disconnect() is called. Since we have 'reader' entry there, StoreMap::closeForReading() is called which just releases the reader lock without marking or freeing the slot. If the entry were 'writer' there (old Squid behavior), StoreMap::abortWriting() would be called, freeing (or marking) the slot.

> And when did the StoreEntry corresponding to the TransientsEntry lost its association with the disk (or memory) store? 

It loses its association with the disk in Rock::SwapDir::disconnect() (called from destroyStoreEntry()) and with memory in MemStore::disconnect(), called via Controller from StoreEntry::destroyMemObject(). That loss seems expected: we disconnect  the StoreEntry from caches before deleting it.

Reply:
> Eventually Transients::disconnect() is called. Since we have 'reader' entry there, StoreMap::closeForReading() is called which just releases the reader lock without marking or freeing the slot.

Aha, this is good news! The underlying "stale entry" problem now boils down to the fact that Transients are reusing a caching-oriented StoreMap which does not release idle valid entries by default. Going forward, I see two options:

1. Add StoreMap::closeForReadingAndFreeIdle() and call it instead of calling closeForReading() in Transients.cc. I see two such closeForReading() calls but my code copy may be stale.
2. Just add a `TODO: s/closeForReading/closeForReadingAndFreeIdle/g` comment to the "we are reading" case of Transients::disconnect().

If it does not have some hidden dangers, I lean towards option 1 because it allows us to get rid of these likely-to-become-stale idle Transients. What do you think?

----

closeForReadingAndFreeIdle() is a long name, but the operation, assuming it is possible to implement at all, has to be atomic from the other workers point of view. Thus, we cannot split it in two. We could add a `freeIfIdle` option to closeForReading() instead, but I think it is best to be more explicit here -- the method code when that option is turned on will probably be rather different from the basic closeForReading() code. If possible, closeForReadingAndFreeIdle() should reuse closeForReading(), of course.

Reply:
> closeForReadingAndFreeIdle() is a long name, but the operation, assuming it is possible to implement at all, has to be atomic from the other workers point of view.

Note that some of the existing StoreMap methods are not 'atomic' when solving similar problems, e.g., StoreMap::abortWriting().  This method checks an atomic condition and performs an action, depending on that condition:

```
if (!s.lock.readers) {
    freeChain(fileno, s, false);
```
If, for example, lock.readers==0 and freeChain() is going to start but another worker has just attached as a reader, it will face a problem, since the slice chain it is going to read is being freed now.

Reply:
> If ... another worker has just attached as a reader

You probably missed the following at the start of StoreMap::abortWriting().

```C++
    s.lock.appending = false; // locks out any new readers
```

abortWriting() is only called for locked-for-writing entries so `lock.appending` is required to allow any concurrent readers.

> some of the existing StoreMap methods are not 'atomic'

IIRC, there are no known bugs in this area.

Reply:
No, StoreMap::abortWriting() seems to be correct: new readers are filtered out, making 'appending' false.

Reply:
Here is a draft of the suggested closeForReadingAndFreeIdle():

```
void
Ipc::StoreMap::closeForReadingAndFreeIdle(const sfileno fileno)
{
    closeForReading(fileno);
    // free if idle
    if (lock.lockExclusive()) {
        assert(!s.writing());
        assert(!s.reading());
        Anchor &s = anchorAt(fileno);
        freeChain(fileno, s, false); 
    }
}

```

Reply:
Is there a way to avoid leaving the "unlocked entry" gap between closeForReading() and lockExclusive() without adding new/complex APIs? I know that the gap does not lead to "bugs", but we should strive for reliable handling of _locked_ entries, and the current draft implementation is not reliable.

Reply:
Done at 96b229f, please review. If we are happy with this solution, I'll undo the fix already applied 
to Controller::allowSharing(), except for that useful 'hittingRequiresCollapsing()' check.

Reply:
> I'll undo the fix already applied to Controller::allowSharing(),

Done at ffc8ea76.